src/circular_buffer.jl

```diff
@@ -1,44 +1,50 @@
-
 """
 New items are pushed to the back of the list, overwriting values in a circular fashion.
 """
-type CircularBuffer{T} <: AbstractVector{T}
+type CircularBuffer{T, N} <: AbstractArray{T, N}
     capacity::Int
     first::Int
-    buffer::Vector{T}
+    buffer::Array{T, N}
+    used::Int
+    colons::Vector{Colon}
+
+    function CircularBuffer(capacity::Int, others_dims...)
+        a = Array{T, N}(capacity, others_dims...)
+        colons = Vector{Colon}(N - 1)
+        new(capacity, 1, a, 0, colons)
+    end
+end
 
-    CircularBuffer(capacity::Int) = new(capacity, 1, T[])
+function CircularBuffer(T, capacity::Int, others_dims...)
+    N = length(others_dims) + 1
+    CircularBuffer{T,N}(capacity::Int, others_dims...)
 end
 
 function _buffer_index(cb::CircularBuffer, i::Int)
     n = length(cb)
     if i < 1 || i > n
         throw(BoundsError("CircularBuffer out of range. cb=$cb i=$i"))
     end
-    idx = cb.first + i - 1
-    if idx > n
-        idx - n
-    else
-        idx
-    end
+    idx = mod1(cb.first + i - 1, n)
 end
 
 function Base.getindex(cb::CircularBuffer, i::Int)
-    cb.buffer[_buffer_index(cb, i)]
+    cb.buffer[_buffer_index(cb, i), cb.colons...]
 end
 
 function Base.setindex!(cb::CircularBuffer, data, i::Int)
-    cb.buffer[_buffer_index(cb, i)] = data
+    cb.buffer[_buffer_index(cb, i), cb.colons...] = data
```

Comment:
I tried to use this code in https://github.com/femtotrader/TimeSeriesIO.jl for a live/stream time series resampler.
see https://github.com/femtotrader/TimeSeriesIO.jl/blob/master/test/test_stream_time_array_ohlcv.jl
and noticed that `setindex!` might be fixed like this:

```julia
function Base.setindex!(cb::CircularBuffer, data, i::Int, args...)
    if length(args) == 0
        colons = cb.colons
    else
        colons = cb.colons[1:length(args)]        
    end
    cb.buffer[_buffer_index(cb, i), args..., colons...] = data
    cb
end
```

Here is my test

```julia
cb = NDCircularBuffer(Int, 3, 4)
#println(convert(Array, cb))
@test size(cb) == (0, 4)
println(cb)
push!(cb, [10, 11, 12, 13])
@test size(cb) == (1, 4)
push!(cb, [20, 21, 22, 23])
@test size(cb) == (2, 4)
push!(cb, [30, 31, 32, 33])
@test size(cb) == (3, 4)
push!(cb, [40, 41, 42, 43])
push!(cb, [50, 51, 52, 53])
@test cb[end-2] == [30,31,32,33]
@test cb[end-1] == [40,41,42,43]
@test cb[end] == [50,51,52,53]
println(cb.buffer)

cb[end, 2] = 99
@test cb[end] == [50,99,52,53]
```

Without this tiny fix `cb[end]` was `[50,51,52,53]` so `setindex!` wasn't working correctly